drivers/modem/gsm_ppp.c

```diff
@@ -399,6 +444,20 @@ static void gsm_finalize_connection(struct gsm_modem *gsm)
 		return;
 	}
 
+wait_reg:
+	/* Wait for cell tower registration */
+	ret = modem_cmd_send_nolock(&gsm->context.iface,
+				    &gsm->context.cmd_handler,
+				    &check_net_reg_cmd, 1,
+				    "AT+CREG?",
+				    &gsm->sem_response,
+				    GSM_CMD_SETUP_TIMEOUT);
+	if ((ret < 0) || ((gsm->net_state != NET_ROAMING) &&
+			 (gsm->net_state != NET_HOME_NETWORK))) {
+		k_sleep(K_SECONDS(1));
+		goto wait_reg;
+	}
```

Comment:
This loop is problematic because the `gsm_finalize_connection()` is called from system workqueue and it should not block very long time. This loop can cause long sleep/delay. Because of this you need to implement this in a state machine instead.

Reply:
Have a look at the Ublox Sara modem driver as it has already addressed this issue.

Reply:
@jukkar Noted, I think I will just have to modify the [wait_reg](https://github.com/zephyrproject-rtos/zephyr/blob/70cb83d7b406e37d6bbf22eaa92c5bafa4affcad/drivers/modem/gsm_ppp.c#L447) part to reschedule the `gsm_configuration_work` in 1 seconds and simply return, then modify the [L402](https://github.com/zephyrproject-rtos/zephyr/blob/70cb83d7b406e37d6bbf22eaa92c5bafa4affcad/drivers/modem/gsm_ppp.c#L402) to jump to `wait_reg` if the `net_state` is 2 (searching). Sounds good to you?

Reply:
@WilliamGFish 
> Have a look at the Ublox Sara modem driver as it has already addressed this issue.

Thanks for your input.
Previously when I was using AT+CREG=2, I tried to do it by waiting for the unsolicited command but somehow it never / took too long to happen so I switched to do AT+CREG? directly instead.

I had a look at the [sara modem](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/ublox-sara-r4.c#L1213) and they are limiting the time by a loop which worst case scenario is 41 seconds delay, I guess this is not preferred as mentioned by @jukkar

Reply:
If you look it:
Attempts to connect
If does not receive response within timeout will reset modem, then retry

If the still unconnected within the 41 secs the driver will restart (ln 1117) 
This will be repeated MDM_NETWORK_RETRY_COUNT times until then giving up and returning an error.

This has been working for a couple of years without issue.

If it can be improved that is great.